### PR TITLE
Add diou_nms for AC

### PR DESCRIPTION
**Pull request by Chen-MingChang**
Because some models, such as yolov4, use DIOU instead of IOU in NMS during training and post-processing.
DIoU summarizes two geometric factors in bounding box regression, i.e., overlap area, and central point distance, thereby leading to faster convergence and better performance.
In this pull, DIoU-NMS function is added to the openVINO accuracy checker tool, so that it can be used to calculate the correct expected accuracy of yolov4.
The contribution of DIoU-NMS can not only collect more accurate accuracy rate, but also can be used with post-training optimization toolkit (pot) to generate better optimized INT8 model.

Taking yolov4 as an example. When using diou_nms, you only need to modify the yml file that makes metric type of postprocessing as diou_nms.
The results show that mAP calculated by diou-nms is more accurate than nms.

yolo-v4-tf :
DIoU-NMS 71.23%
NMS 71.17%

### CODE REVIEW

**Comment by vladimir-dudnik** [comment]
Jenkins please retry a build

File: models/public/yolo-v4-tf/accuracy-check.yml [hunk 1]
```diff
@@ -41,7 +41,7 @@ models:
             apply_to: prediction
             min_confidence: 0.001
             remove_filtered: true
-          - type: nms
+          - type: diou_nms
```
Comment (eaidova) [inline models/public/yolo-v4-tf/accuracy-check.yml:44]: @eizamaliev could you please comment, why during adding yolo-v4 you do not use diou_nms?
@playboy007gy I want to note that we do not have intention to get the most accurate results that possible, our purposes to reproduce original testing condition and as I understand this posprocessing used only during training, final result in the article provided without it.

File: tools/accuracy_checker/accuracy_checker/postprocessor/nms.py [hunk 2]
```diff
@@ -230,3 +230,93 @@ def _nms(self, input_bboxes, input_scores):
...
+        parameters.update({
+            'overlap': NumberField(
+                min_value=0, max_value=1, optional=True, default=0.5,
+                description="Overlap threshold for merging detections."
+            ),
+            'include_boundaries': BoolField(
+                optional=True, default=True, description="Shows if boundaries are included."
+            ),
+            'keep_top_k': NumberField(min_value=0, optional=True, description="Keep top K."),
+            'use_min_area': BoolField(
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/postprocessor/nms.py:252]: please remove unused parameter

File: tools/accuracy_checker/accuracy_checker/postprocessor/nms.py [hunk 3]
```diff
@@ -230,3 +230,93 @@ def _nms(self, input_bboxes, input_scores):
...
+            )
+            prediction.remove([box for box in range(len(prediction.x_mins)) if box not in keep])
+
+        return annotations, predictions
+
+    @staticmethod
+    def diou_nms(x1, y1, x2, y2, scores, thresh, include_boundaries=True, keep_top_k=None, use_min_area=False):
+        """
+        Pure Python NMS baseline.
+        """
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/postprocessor/nms.py:280]: please remove comment, which is not related to your method

File: tools/accuracy_checker/accuracy_checker/postprocessor/nms.py [hunk 4]
```diff
@@ -230,3 +230,93 @@ def _nms(self, input_bboxes, input_scores):
...
+                base_area = np.minimum(areas[i], areas[order[1:]])
+            else:
+                base_area = (areas[i] + areas[order[1:]] - intersection)
+
+            overlap = np.divide(
+                intersection,
+                base_area,
+                out=np.zeros_like(intersection, dtype=float),
+                where=base_area != 0
+            ) - pow(rh02/c_area,0.6)
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/postprocessor/nms.py:319]:
```suggestion
            ) - pow(rh02 / c_area, 0.6)
```

File: tools/accuracy_checker/accuracy_checker/postprocessor/nms.py [hunk 5]
```diff
@@ -230,3 +230,93 @@ def _nms(self, input_bboxes, input_scores):
...
+            yy2 = np.minimum(y2[i], y2[order[1:]])
+
+            w = np.maximum(0.0, xx2 - xx1 + b)
+            h = np.maximum(0.0, yy2 - yy1 + b)
+            intersection = w * h
+
+            cw = np.maximum(x2[i], x2[order[1:]]) - np.minimum(x1[i], x1[order[1:]])
+            ch = np.maximum(y2[i], y2[order[1:]]) - np.minimum(y1[i], y1[order[1:]])
+            c_area = cw**2+ch**2+1e-16
+            rh02 = ((x2[order[1:]]+x1[order[1:]])-(x2[i]+x1[i]))**2/4+((y2[order[1:]]+y1[order[1:]])-(y2[i]+y1[i]))**2/4
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/postprocessor/nms.py:307]:
```suggestion
            rh02 = ((x2[order[1:]] + x1[order[1:]]) - (x2[i] + x1[i]))**2 / 4+((y2[order[1:]] + y1[order[1:]]) -(y2[i] + y1[i]))**2 / 4
```

File: tools/accuracy_checker/accuracy_checker/postprocessor/nms.py (hunk 2 above)
Comment (playboy007gy) [reply tools/accuracy_checker/accuracy_checker/postprocessor/nms.py:252]: Which parts you mean? 
I think these parameters are used in 

def process_image(self, annotations, predictions):
        for prediction in predictions:
            scores = get_scores(prediction)
            keep = self.diou_nms(
                prediction.x_mins, prediction.y_mins, prediction.x_maxs, prediction.y_maxs, scores,
                self.overlap, self.include_boundaries, self.keep_top_k, self.use_min_area
            )
            prediction.remove([box for box in range(len(prediction.x_mins)) if box not in keep])

        return annotations, predictions

Comment (eaidova) [reply tools/accuracy_checker/accuracy_checker/postprocessor/nms.py:252]: `use_min_area` is extension of NMS algo, which is not used in diouNMS, so I think you do not need it

Comment (playboy007gy) [reply tools/accuracy_checker/accuracy_checker/postprocessor/nms.py]: You are right, I have removed 'use_min_area'.

File: tools/accuracy_checker/accuracy_checker/postprocessor/nms.py (hunk 3 above)
Comment (playboy007gy) [reply tools/accuracy_checker/accuracy_checker/postprocessor/nms.py]: done

File: tools/accuracy_checker/accuracy_checker/postprocessor/nms.py (hunk 4 above)
Comment (playboy007gy) [reply tools/accuracy_checker/accuracy_checker/postprocessor/nms.py]: done

File: tools/accuracy_checker/accuracy_checker/postprocessor/nms.py (hunk 5 above)
Comment (playboy007gy) [reply tools/accuracy_checker/accuracy_checker/postprocessor/nms.py]: done

File: models/public/yolo-v4-tf/accuracy-check.yml (hunk 1 above)
Comment (playboy007gy) [reply models/public/yolo-v4-tf/accuracy-check.yml:44]: In the original paper, diou was used for training and inference.
(The purpose of inference is to show the ability gained from training, so the same algorithm as the training condition must be used)
And we want to reproduce the original testing conditions, I think diou is also necessary.

**Comment by eaidova** [comment]
Jenkins ok to test

**Comment by eaidova** [comment]
@playboy007gy please also update supported postprocessing docs:
https://github.com/openvinotoolkit/open_model_zoo/blob/develop/tools/accuracy_checker/accuracy_checker/postprocessor/README.md

and register your class in postprocessing module:
https://github.com/openvinotoolkit/open_model_zoo/blob/develop/tools/accuracy_checker/accuracy_checker/postprocessor/__init__.py

**Comment by eaidova** [comment]
@playboy007gy testing of your changes failed with syntax error. Could you please fix?
```
11:05:16  ./tools/accuracy_checker/accuracy_checker/postprocessor/nms.py:1:1: E902 IndentationError: unindent does not match any outer indentation level
11:05:16  ./tools/accuracy_checker/accuracy_checker/postprocessor/nms.py:252:13: E123 closing bracket does not match indentation of opening bracket's line
11:05:16  ./tools/accuracy_checker/accuracy_checker/postprocessor/nms.py:265:17: E113 unexpected indentation

11:06:29  E     File "/home/jenkins/workspace/omz/precommit/accuracy_checker_ubuntu18/src/tools/accuracy_checker/accuracy_checker/postprocessor/nms.py", line 252
11:06:29  E       )
11:06:29  E       ^
11:06:29  E   SyntaxError: invalid syntax
```

File: tools/accuracy_checker/accuracy_checker/postprocessor/nms.py [hunk 6]
```diff
@@ -230,3 +230,84 @@ def _nms(self, input_bboxes, input_scores):
             scores *= np.exp(np.negative(np.square(iou_values) / self.sigma))
 
         return np.array(out_ids, dtype=np.int32), np.array(out_scores, dtype=np.float32)
+
+class DIOU_NMS(Postprocessor):
```
Comment (eaidova) [inline tools/accuracy_checker/accuracy_checker/postprocessor/nms.py:234]: naming convention does not allow to use underscore in class names
```suggestion
class DIoUNMS(Postprocessor):
```
or
```suggestion
class DistanceIoUNMS(Postprocessor):
```

**Comment by vladimir-dudnik** [comment]
please fix precommit check

> 11:28:48  ************* Module accuracy_checker.postprocessor.nms
11:28:48  accuracy_checker/postprocessor/nms.py:300: [C0301(line-too-long), ] Line too long (136/120)
11:28:48  accuracy_checker/postprocessor/nms.py:309: [C0326(bad-whitespace), ] Exactly one space required after comma
11:28:48              ) - pow(rh02 / c_area,0.6)

File: models/public/yolo-v4-tf/accuracy-check.yml (hunk 1 above)
Comment (eizamaliev) [reply models/public/yolo-v4-tf/accuracy-check.yml:44]: @eaidova Yes, indeed. They add diou nms sorting in [commint](https://github.com/AlexeyAB/darknet/commit/c7e3e2ee9ec2d8fff447d83736e13bce0938015f).

**Review by eaidova — APPROVED** [review]
from implementation point of view, it is nice to have DIoU NMS
@Chen-MingChang thank you,
but still not sure that we should add it in configuration files.

**Pull request merged**